packages/unified-selection/src/unified-selection/iModel/IModelSelectionHandler.ts

```diff
@@ -0,0 +1,221 @@
+        },
+        complete: () => {
+          this.applyCurrentHiliteSet(false);
+        },
+      });
+  }
+
+  private onUnifiedSelectionChanged = (args: StorageSelectionChangeEventArgs) => {
+    // iModels are only interested in top-level selection changes
+    if (args.iModelKey !== this._iModel.key || args.level !== 0) {
+      return;
+    }
+
+    this._cancelOngoingChanges.next();
+    this.handleUnifiedSelectionChange(args.changeType, args.selectables);
+  };
+
+  private applyCurrentHiliteSet(clearBefore = true) {
+    if (clearBefore) {
+      using(this.suspendIModelToolSelectionSync(), (_) => {
+        this._iModel.hilited.clear();
+        this._iModel.selectionSet.emptyAll();
+      });
+    }
+
+    from(this._cachingHiliteSetProvider.getHiliteSet({ iModelKey: this._iModel.key }))
+      .pipe(takeUntil(this._cancelOngoingChanges))
+      .subscribe({
+        next: (ids) => {
+          this.applyHiliteSet(ids);
+        },
+      });
+  }
+
+  private applyHiliteSet(set: HiliteSet) {
+    using(this.suspendIModelToolSelectionSync(), (_) => {
+      if (set.models && set.models.length) {
+        this._iModel.hilited.models.addIds(set.models);
+      }
+      if (set.subCategories && set.subCategories.length) {
+        this._iModel.hilited.subcategories.addIds(set.subCategories);
+      }
+      if (set.elements && set.elements.length) {
+        this._iModel.hilited.elements.addIds(set.elements);
+        this._iModel.selectionSet.add(set.elements);
+      }
+    });
+  }
+
+  private removeHiliteSet(set: HiliteSet) {
+    using(this.suspendIModelToolSelectionSync(), (_) => {
+      if (set.models.length) {
+        this._iModel.hilited.models.deleteIds(set.models);
+      }
+      if (set.subCategories.length) {
+        this._iModel.hilited.subcategories.deleteIds(set.subCategories);
+      }
+      if (set.elements.length) {
+        this._iModel.hilited.elements.deleteIds(set.elements);
+        this._iModel.selectionSet.remove(set.elements);
+      }
+    });
+  }
+
+  private onIModelSelectionChanged = async (event: SelectionSetEvent): Promise<void> => {
+    if (this._isSuspended || event.set.iModel !== this._iModel) {
+      return;
+    }
+
+    if (SelectionSetEventType.Clear === event.type) {
+      this._selectionStorage.clearSelection({ iModelKey: this._iModel.key, source: this._selectionSourceName });
+      return;
+    }
+
+    const elementIds = this.getSelectionSetChangeIds(event);
+    const scopedSelection = computeSelection({ queryExecutor: this._queryExecutor, elementIds, scope: this._activeScopeProvider() });
+    await this.handleIModelSelectionChange(event.type, scopedSelection);
+  };
+
+  private async handleIModelSelectionChange(type: SelectionSetEventType, iterator: AsyncIterableIterator<SelectableInstanceKey>) {
+    if (type === SelectionSetEventType.Replace) {
+      this._selectionStorage.clearSelection({ iModelKey: this._iModel.key, source: this._selectionSourceName });
+    }
```

Comment:
Can we move this under handling of `Remove` to get it closed to the other part of `Replace` handling?

Reply:
Moved it closer to the add part.